### PR TITLE
refactor: add a FC layer after RNN

### PR DESCRIPTION
**Pull request by kkweon**
1. Add a FC Layer after RNN
  - #73
2. Add a numpy doc style
3. Explicitly define an activation function in `tf.contrib.rnn.BasicLSTMCell`
  - default is `tf.tanh`
  - [Tensorflow official API link](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/BasicLSTMCell)
4. Change to the `AdamOptimizer` from the SGD
5. Use `with` statement. Otherwise, the session must be closed explicitly
6. No linebreak for the 79 char limit

### CODE REVIEW

File: lab-12-5-rnn_stock_prediction.py [hunk 1]
```diff
@@ -31,47 +58,49 @@ def MinMaxScaler(data):
...
 Y = tf.placeholder(tf.float32, [None, 1])
 
-cell = tf.contrib.rnn.BasicLSTMCell(num_units=output_dim, state_is_tuple=True)
+# build a LSTM network
+cell = tf.contrib.rnn.BasicLSTMCell(num_units=hidden_dim, state_is_tuple=True, activation=tf.tanh)
 outputs, _states = tf.nn.dynamic_rnn(cell, X, dtype=tf.float32)
-Y_pred = outputs[:, -1]  # We use the last cell's output
-
-print(outputs[:, -1])
+Y_pred = tf.contrib.layers.fully_connected(outputs[:, -1], output_dim, activation_fn=None)  # We use the last cell's output
```
Comment (hunkim) [inline lab-12-5-rnn_stock_prediction.py]: Do you think it's a good idea to introduce tf.contrib.layers.fully_connected? Or just simple WX+b like previous examples?

File: lab-12-5-rnn_stock_prediction.py [hunk 2]
```diff
@@ -31,47 +58,49 @@ def MinMaxScaler(data):
...
+trainX, testX = np.array(dataX[0:train_size]), np.array(dataX[train_size:len(dataX)])
+trainY, testY = np.array(dataY[0:train_size]), np.array(dataY[train_size:len(dataY)])
 
 # input place holders
 X = tf.placeholder(tf.float32, [None, seq_length, data_dim])
 Y = tf.placeholder(tf.float32, [None, 1])
 
-cell = tf.contrib.rnn.BasicLSTMCell(num_units=output_dim, state_is_tuple=True)
+# build a LSTM network
+cell = tf.contrib.rnn.BasicLSTMCell(num_units=hidden_dim, state_is_tuple=True, activation=tf.tanh)
```
Comment (hunkim) [inline lab-12-5-rnn_stock_prediction.py]: Do you think we can set activation=None? Then, perhaps we don't need a FC after RNN.

Comment (kkweon) [inline lab-12-5-rnn_stock_prediction.py]: I think it's a good point to introduce `tf.contrib.layers` because RNN is also from the contrib..

File: lab-12-5-rnn_stock_prediction.py (hunk 1 above)
Comment (kkweon) [inline lab-12-5-rnn_stock_prediction.py]: I think it's a good point to introduce tf.contrib.layers because RNN is also from the contrib..

File: lab-12-5-rnn_stock_prediction.py (hunk 2 above)
Comment (kkweon) [inline lab-12-5-rnn_stock_prediction.py]: I have a feeling it's not correct..

image

Because even if we remove `tanh`, there is an output gate with the sigmoid

😃

Comment (kkweon) [inline lab-12-5-rnn_stock_prediction.py]: Yup I tried. it threw out errors. It requires the activation function.

**Comment by hunkim** [comment]
How about vanilla RNN (no LSTM)?

**Comment by kkweon** [comment]
Same. a Vanilla RNN still requires an activation function.

**Comment by hunkim** [comment]
How about `tf.nn.bias_add` activation function (with 0 bias)? https://www.tensorflow.org/versions/r0.10/api_docs/python/nn/activation_functions_#softplus

**Comment by kkweon** [comment]
then we have to use a closure like this
```
none = lambda x: tf.nn.bias_add(x, [0])
cell = tf.contrib.rnn.BasicRNNCell(num_units=hidden_dim, activation=none)
```

this will work though I am not sure if it's okay to do...

we are repeating the following 7 times.
h = Wxh * x  + Whh * h
where h is an output at each step.

**Comment by hunkim** [comment]
This is not pretty. Let's just add one FC layer.

Sung
